File: hphp/hack/src/typing/typing_make_type.ml
(*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

open Typing_defs
open Typing_defs_constraints
module SN = Naming_special_names
module Reason = Typing_reason
module Nast = Aast

let class_type r name tyl =
  mk (r, Tclass ((Reason.to_pos r, name), nonexact, tyl))

let prim_type r t = mk (r, Tprim t)

(* Make a negation type *)
let neg r neg_t =
  match snd neg_t with
  (* Represent the negation of Tnull as Tnonnull, instead of Tneg Tnull *)
  | IsTag NullTag -> mk (r, Tnonnull)
  | _ -> mk (r, Tneg neg_t)

let traversable r ty = class_type r SN.Collections.cTraversable [ty]

let keyed_traversable r kty vty =
  class_type r SN.Collections.cKeyedTraversable [kty; vty]

let keyed_container r kty vty =
  class_type r SN.Collections.cKeyedContainer [kty; vty]

let awaitable r ty = class_type r SN.Classes.cAwaitable [ty]

let generator r key value send =
  class_type r SN.Classes.cGenerator [key; value; send]

let async_generator r key value send =
  class_type r SN.Classes.cAsyncGenerator [key; value; send]

let async_iterator r ty = class_type r SN.Classes.cAsyncIterator [ty]

let async_keyed_iterator r ty1 ty2 =
  class_type r SN.Classes.cAsyncKeyedIterator [ty1; ty2]

let pair r ty1 ty2 = class_type r SN.Collections.cPair [ty1; ty2]

let dict r kty vty = class_type r SN.Collections.cDict [kty; vty]

let keyset r ty = class_type r SN.Collections.cKeyset [ty]

let vec r ty = class_type r SN.Collections.cVec [ty]

let any_array r kty vty = class_type r SN.Collections.cAnyArray [kty; vty]

let container r ty = class_type r SN.Collections.cContainer [ty]

let throwable r = class_type r SN.Classes.cThrowable []

let datetime r = class_type r SN.Classes.cDateTime []

let datetime_immutable r = class_type r SN.Classes.cDateTimeImmutable []

let const_vector r ty = class_type r SN.Collections.cConstVector [ty]

let const_collection r ty = class_type r SN.Collections.cConstCollection [ty]

let collection r ty = class_type r SN.Collections.cCollection [ty]

let spliceable r ty1 ty2 ty3 =
  class_type r SN.Classes.cSpliceable [ty1; ty2; ty3]

let vec_or_dict r kty vty = mk (r, Tvec_or_dict (kty, vty))

let int r = prim_type r Nast.Tint

let bool r = prim_type r Nast.Tbool

let string r = class_type r SN.Classes.cString []

(* Mirror of the classname.hhi definition. This sucks: the type of the ::class
 * constant is burned into decls at folding time so this supports the case when
 * you want to wrap a locl ty e.g. from [Typing.class_expr] in a classname.
 * TODO: eliminate the use site of this for CIexpr and make class_expr
 * produce a class type directly *)
let typename r tyl = mk (r, Tnewtype (SN.Classes.cTypename, tyl, string r))

let classname r tyl =
  mk (r, Tnewtype (SN.Classes.cClassname, tyl, typename r tyl))

let float r = prim_type r Nast.Tfloat

let num r = prim_type r Nast.Tnum

let arraykey r = prim_type r Nast.Tarraykey

let void (type ph) r : ph ty = prim_type r Nast.Tvoid

let null r = prim_type r Nast.Tnull

let nonnull r = mk (r, Tnonnull)

let dynamic r = mk (r, Tdynamic)

let is_dynamic_or_like_or_mixed ty =
  match get_node ty with
  | Tdynamic
  | Tlike _
  | Tmixed ->
    true
  | _ -> false

let like r ty =
  if is_dynamic_or_like_or_mixed ty then
    ty
  else
    match get_node ty with
    | Tapply ((_, n), [inner_ty])
      when String.equal n Naming_special_names.Classes.cSupportDyn
           && is_dynamic_or_like_or_mixed inner_ty ->
      ty
    | _ -> mk (r, Tlike ty)

let locl_like r ty =
  if Typing_defs.is_dynamic ty then
    ty
  else
    match get_node ty with
    | Tprim Aast.Tnoreturn
    | Tany _ ->
      ty
    | Tunion tys when List.exists Typing_defs.is_dynamic tys -> ty
    | _ -> mk (r, Tunion [dynamic r; ty])

(* Wrap supportdyn around a type unless it's already got it
 * or trivially supports dynamic
 *)
let supportdyn r ty =
  match get_node ty with
  | Tnewtype (c, _, _) when String.equal c SN.Classes.cSupportDyn -> ty
  | Tunion []
  | Tdynamic
  | Tprim _ ->
    ty
  | _ -> mk (r, Tnewtype (SN.Classes.cSupportDyn, [ty], ty))

let supportdyn_nonnull r = supportdyn r (nonnull r)

let mixed r = mk (r, Toption (nonnull r))

let nothing r = mk (r, Tunion [])

let shape r kind map =
  mk
    ( r,
      Tshape
        { s_origin = Missing_origin; s_unknown_value = kind; s_fields = map } )

let closed_shape r map = shape r (nothing r) map

let open_shape ~kind r map = shape r kind map

let supportdyn_mixed r = supportdyn r (mixed r)

let resource r = prim_type r Nast.Tresource

let tyvar r v = mk (r, Tvar v)

let generic r n = mk (r, Tgeneric n)

let this r = mk (r, Tgeneric SN.Typehints.this)

let taccess r ty id = mk (r, Taccess (ty, id))

let nullable : type a. a Reason.t_ -> a ty -> a ty =
 fun r ty ->
  (* Cheap avoidance of double nullable *)
  match get_node ty with
  | Toption _ as ty_ -> mk (r, ty_)
  | Tunion [] -> null r
  | _ -> mk (r, Toption ty)

let apply r id tyl = mk (r, Tapply (id, tyl))

let decl_string r = apply r (Reason.to_pos r, SN.Classes.cString) []

let tuple r tyl =
  mk
    ( r,
      Ttuple
        { t_required = tyl; t_optional = []; t_extra = Tvariadic (nothing r) }
    )

let union r tyl =
  match tyl with
  | [ty] -> ty
  | _ -> mk (r, Tunion tyl)

let intersection r tyl =
  match tyl with
  | [] -> mixed r
  | [ty] -> ty
  | _ -> mk (r, Tintersection tyl)

let function_ref r ty = mk (r, Tnewtype (SN.Classes.cFunctionRef, [ty], ty))

let label r ty_in ty_out =
  mk (r, Tnewtype (SN.Classes.cEnumClassLabel, [ty_in; ty_out], mixed r))

let has_member r ~name ~ty ~class_id ~methd =
  ConstraintType
    (mk_constraint_type
       ( r,
         Thas_member
           {
             hm_name = name;
             hm_type = ty;
             hm_class_id = class_id;
             hm_method = methd;
           } ))

let list_destructure r tyl =
  ConstraintType
    (mk_constraint_type
       ( r,
         Tdestructure
           {
             d_required = tyl;
             d_optional = [];
             d_variadic = None;
             d_kind = ListDestructure;
           } ))

let simple_variadic_splat r ty =
  ConstraintType
    (mk_constraint_type
       ( r,
         Tdestructure
           {
             d_required = [];
             d_optional = [];
             d_variadic = Some ty;
             d_kind = SplatUnpack;
           } ))

let capability r name : locl_ty = class_type r name []

let default_capability p : locl_ty =
  let r = Reason.default_capability p in
  intersection
    r
    SN.Capabilities.
      [
        class_type r writeProperty [];
        class_type r accessGlobals [];
        class_type r rxLocal [];
        class_type r systemLocal [];
        class_type r implicitPolicyLocal [];
        class_type r io [];
      ]

let default_capability_decl p : decl_ty =
  let r = Reason.default_capability p in
  intersection
    r
    SN.Capabilities.
      [
        apply r (p, writeProperty) [];
        apply r (p, accessGlobals) [];
        apply r (p, rxLocal) [];
        apply r (p, systemLocal) [];
        apply r (p, implicitPolicyLocal) [];
        apply r (p, io) [];
      ]

let default_capability_unsafe p : locl_ty = mixed (Reason.hint p)

(** Construct the default type for `__construct` *)
let default_construct (type ph) r : ph ty =
  mk
    ( r,
      Tfun
        {
          ft_tparams = [];
          ft_where_constraints = [];
          ft_params = [];
          ft_implicit_params = { capability = CapDefaults (Reason.to_pos r) };
          ft_flags =
            Typing_defs_flags.Fun.make
              Ast_defs.FSync
              ~return_disposable:false
              ~returns_readonly:false
              ~readonly_this:false
              ~support_dynamic_type:false
              ~is_memoized:false
              ~variadic:false;
          ft_ret = void r;
          ft_instantiated = true;
        } )

let make_dynamic_tfun r : locl_ty =
  mk
    ( r,
      Tfun
        {
          ft_tparams = [];
          ft_where_constraints = [];
          ft_params =
            [
              {
                fp_pos = Reason.to_pos r;
                fp_name = None;
                fp_type = dynamic r;
                fp_flags =
                  Typing_defs_flags.FunParam.make
                    ~inout:false
                    ~accept_disposable:false
                    ~is_optional:false
                    ~readonly:false
                    ~ignore_readonly_error:false
                    ~splat:true
                    ~named:false;
                fp_def_value = None;
              };
            ];
          ft_implicit_params = { capability = CapDefaults (Reason.to_pos r) };
          ft_flags =
            Typing_defs_flags.Fun.make
              Ast_defs.FSync
              ~return_disposable:false
              ~returns_readonly:false
              ~readonly_this:false
              ~support_dynamic_type:false
              ~is_memoized:false
              ~variadic:true;
          ft_ret = dynamic r;
          ft_instantiated = true;
        } )

File: hphp/hack/src/typing/typing_make_type.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

use bumpalo::collections::String as BString;
use bumpalo::collections::Vec as BVec;
use bumpalo::{vec, Bump};

use arena_trait::Arena;
use naming_special_names_rust::{classes, collections};
use oxidized_by_ref::pos::Pos;
use oxidized_by_ref::tany_sentinel::TanySentinel;
use oxidized_by_ref::{
    ident,
    typing_defs_flags::{FunParamFlags, FunTypeFlags},
};

use crate::typing_defs::ExpandEnv;
use crate::typing_defs_core::*;
use crate::typing_logic::{SubtypeProp, SubtypePropEnum};
use crate::typing_reason::*;
use typing_collections_rust::{SMap, Vec};

// Struct off which we call type builder methods
// This gives us the option to keep some state here e.g. for hash consing
pub struct TypeBuilder<'a> {
    // Well-known identifiers
    // We could initialize the positions to their actual definition site in hhi
    // This makes more sense than inheriting the position from the use
    // But for now, let's just put none as the position
    // TODO: we might consider creating a SymbolBuilder type
    id_traversable: PosId<'a>,
    id_keyed_traversable: PosId<'a>,
    id_keyed_container: PosId<'a>,
    id_awaitable: PosId<'a>,
    id_generator: PosId<'a>,
    id_async_generator: PosId<'a>,
    id_async_iterator: PosId<'a>,
    id_async_keyed_iterator: PosId<'a>,
    id_pair: PosId<'a>,
    id_dict: PosId<'a>,
    id_keyset: PosId<'a>,
    id_vec: PosId<'a>,
    id_container: PosId<'a>,
    id_throwable: PosId<'a>,
    id_datetime: PosId<'a>,
    id_datetime_immutable: PosId<'a>,
    id_const_vector: PosId<'a>,
    id_const_collection: PosId<'a>,
    id_collection: PosId<'a>,

    /// The allocator.
    ///
    /// We have made this field private to avoid people calling
    /// alloc on the underlying allocator directly. Instead, use
    /// `TypeBuilder::alloc`, which provides safety mechanisms
    /// against memory leaks.
    bumpalo: &'a Bump,
}

impl<'a> Arena for TypeBuilder<'a> {
    #[inline(always)]
    fn alloc<T>(&self, val: T) -> &mut T {
        return self.bumpalo.alloc(val);
    }
}

// Copying of string is bad
// Perhaps best is to actually *resolve* well-known classes
// as though they appear in the source, so we get the position
// from the definition e.g. in hhi file.
fn mk_special_id(name: &'static str) -> PosId<'static> {
    (Pos::none(), name)
}

impl<'a> TypeBuilder<'a> {
    pub fn new(bumpalo: &'a Bump) -> Self {
        TypeBuilder {
            bumpalo,

            id_traversable: mk_special_id(collections::TRAVERSABLE),
            id_keyed_traversable: mk_special_id(collections::KEYED_TRAVERSABLE),
            id_keyed_container: mk_special_id(collections::KEYED_CONTAINER),
            id_awaitable: mk_special_id(classes::AWAITABLE),
            id_generator: mk_special_id(classes::GENERATOR),
            id_async_generator: mk_special_id(classes::ASYNC_GENERATOR),
            id_async_iterator: mk_special_id(classes::ASYNC_ITERATOR),
            id_async_keyed_iterator: mk_special_id(classes::ASYNC_KEYED_ITERATOR),
            id_pair: mk_special_id(collections::PAIR),
            id_dict: mk_special_id(collections::DICT),
            id_keyset: mk_special_id(collections::KEYSET),
            id_vec: mk_special_id(collections::VEC),
            id_container: mk_special_id(collections::CONTAINER),
            id_throwable: mk_special_id(classes::THROWABLE),
            id_datetime: mk_special_id(classes::DATE_TIME),
            id_datetime_immutable: mk_special_id(classes::DATE_TIME_IMMUTABLE),
            id_const_vector: mk_special_id(collections::CONST_VECTOR),
            id_const_collection: mk_special_id(collections::CONST_COLLECTION),
            id_collection: mk_special_id(collections::COLLECTION),
        }
    }

    pub fn bumpalo(&self) -> &'a Bump {
        self.bumpalo
    }

    pub fn vec_from_iter<T, I: IntoIterator<Item = T>>(&self, iter: I) -> BVec<'a, T> {
        BVec::from_iter_in(iter, &self.bumpalo)
    }

    pub fn slice_from_iter<T, I: IntoIterator<Item = T>>(&self, iter: I) -> &'a [T] {
        self.vec_from_iter(iter).into_bump_slice()
    }

    pub fn str_from_str(&self, s: &str) -> &'a str {
        BString::from_str_in(s, &self.bumpalo).into_bump_str()
    }
}

/// All type builders go here
impl<'a> TypeBuilder<'a> {
    // All type construction should go through here
    fn mk(&'a self, reason: &'a Reason<'a>, ty_: Ty_<'a>) -> &'a Ty<'a> {
        self.alloc(Ty::mk(reason, ty_))
    }

    pub fn any(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.mk(reason, Ty_::Tany(TanySentinel))
    }
    pub fn prim(&'a self, reason: &'a Reason<'a>, kind: PrimKind) -> &'a Ty<'a> {
        self.mk(reason, Ty_::Tprim(self.alloc(kind)))
    }

    pub fn shape(
        &'a self,
        reason: &'a Reason<'a>,
        kind: ShapeKind,
        map: t_shape_map::TShapeMap<'a, &'a ShapeFieldType<'a>>,
    ) -> &'a Ty<'a> {
        self.mk(reason, Ty_::Tshape(self.alloc((kind, map))))
    }

    pub fn class(
        &'a self,
        reason: &'a Reason<'a>,
        name: PosId<'a>,
        tys: &'a [&'a Ty<'a>],
    ) -> &'a Ty<'a> {
        self.mk(
            reason,
            Ty_::Tclass(self.alloc((name, Exact::Nonexact, tys))),
        )
    }
    pub fn traversable(&'a self, reason: &'a Reason<'a>, ty: &'a Ty<'a>) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_traversable,
            vec![in &self.bumpalo; ty].into_bump_slice(),
        )
    }
    pub fn keyed_traversable(
        &'a self,
        reason: &'a Reason<'a>,
        kty: &'a Ty<'a>,
        vty: &'a Ty<'a>,
    ) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_keyed_traversable,
            vec![in &self.bumpalo; kty, vty].into_bump_slice(),
        )
    }
    pub fn keyed_container(
        &'a self,
        reason: &'a Reason<'a>,
        kty: &'a Ty<'a>,
        vty: &'a Ty<'a>,
    ) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_keyed_container,
            vec![in &self.bumpalo; kty, vty].into_bump_slice(),
        )
    }
    pub fn awaitable(&'a self, reason: &'a Reason<'a>, ty: &'a Ty<'a>) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_awaitable,
            vec![in &self.bumpalo; ty].into_bump_slice(),
        )
    }
    pub fn generator(
        &'a self,
        reason: &'a Reason<'a>,
        key: &'a Ty<'a>,
        value: &'a Ty<'a>,
        send: &'a Ty<'a>,
    ) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_generator,
            vec![in &self.bumpalo; key, value, send].into_bump_slice(),
        )
    }
    pub fn async_generator(
        &'a self,
        reason: &'a Reason<'a>,
        key: &'a Ty<'a>,
        value: &'a Ty<'a>,
        send: &'a Ty<'a>,
    ) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_async_generator,
            vec![in &self.bumpalo; key, value, send].into_bump_slice(),
        )
    }
    pub fn async_iterator(&'a self, reason: &'a Reason<'a>, ty: &'a Ty<'a>) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_async_iterator,
            vec![in &self.bumpalo; ty].into_bump_slice(),
        )
    }
    pub fn async_keyed_iterator(
        &'a self,
        reason: &'a Reason<'a>,
        kty: &'a Ty<'a>,
        vty: &'a Ty<'a>,
    ) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_async_keyed_iterator,
            vec![in &self.bumpalo; kty, vty].into_bump_slice(),
        )
    }
    pub fn pair(&'a self, reason: &'a Reason<'a>, ty1: &'a Ty<'a>, ty2: &'a Ty<'a>) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_pair,
            vec![in &self.bumpalo; ty1, ty2].into_bump_slice(),
        )
    }
    pub fn dict(&'a self, reason: &'a Reason<'a>, kty: &'a Ty<'a>, vty: &'a Ty<'a>) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_dict,
            vec![in &self.bumpalo; kty, vty].into_bump_slice(),
        )
    }
    pub fn keyset(&'a self, reason: &'a Reason<'a>, ty: &'a Ty<'a>) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_keyset,
            vec![in &self.bumpalo; ty].into_bump_slice(),
        )
    }
    pub fn vec(&'a self, reason: &'a Reason<'a>, ty: &'a Ty<'a>) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_vec,
            vec![in &self.bumpalo; ty].into_bump_slice(),
        )
    }
    pub fn container(&'a self, reason: &'a Reason<'a>, ty: &'a Ty<'a>) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_container,
            vec![in &self.bumpalo; ty].into_bump_slice(),
        )
    }
    pub fn const_vector(&'a self, reason: &'a Reason<'a>, ty: &'a Ty<'a>) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_const_vector,
            vec![in &self.bumpalo; ty].into_bump_slice(),
        )
    }
    pub fn const_collection(&'a self, reason: &'a Reason<'a>, ty: &'a Ty<'a>) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_const_collection,
            vec![in &self.bumpalo; ty].into_bump_slice(),
        )
    }
    pub fn collection(&'a self, reason: &'a Reason<'a>, ty: &'a Ty<'a>) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_collection,
            vec![in &self.bumpalo; ty].into_bump_slice(),
        )
    }
    pub fn throwable(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_throwable,
            vec![in &self.bumpalo; ].into_bump_slice(),
        )
    }
    pub fn datetime(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_datetime,
            vec![in &self.bumpalo].into_bump_slice(),
        )
    }
    pub fn datetime_immutable(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.class(
            reason,
            self.id_datetime_immutable,
            vec![in &self.bumpalo].into_bump_slice(),
        )
    }
    pub fn int(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.prim(reason, PrimKind::Tint)
    }
    pub fn bool(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.prim(reason, PrimKind::Tbool)
    }
    pub fn string(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.prim(reason, PrimKind::Tstring)
    }
    pub fn float(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.prim(reason, PrimKind::Tfloat)
    }
    pub fn arraykey(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.prim(reason, PrimKind::Tarraykey)
    }
    pub fn void(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.prim(reason, PrimKind::Tvoid)
    }
    pub fn null(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.prim(reason, PrimKind::Tnull)
    }
    pub fn nonnull(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.mk(reason, Ty_::Tnonnull)
    }
    pub fn dynamic(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.mk(reason, Ty_::Tdynamic)
    }
    /*
    pub fn object(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.mk(reason, Ty_::Tobject)
    }
    */
    pub fn tyvar(&'a self, reason: &'a Reason<'a>, v: ident::Ident) -> &'a Ty<'a> {
        self.mk(reason, Ty_::Tvar(v))
    }
    pub fn union(&'a self, reason: &'a Reason<'a>, tys: &'a [&'a Ty<'a>]) -> &'a Ty<'a> {
        self.mk(reason, Ty_::Tunion(tys))
    }
    pub fn intersection(&'a self, reason: &'a Reason<'a>, tys: &'a [&'a Ty<'a>]) -> &'a Ty<'a> {
        self.mk(reason, Ty_::Tintersection(tys))
    }
    pub fn nothing(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.union(reason, &[])
    }
    pub fn nullable(&'a self, reason: &'a Reason<'a>, ty: &'a Ty<'a>) -> &'a Ty<'a> {
        self.mk(reason, Ty_::Toption(ty))
    }
    pub fn mixed(&'a self, reason: &'a Reason<'a>) -> &'a Ty<'a> {
        self.nullable(reason, self.nonnull(reason))
    }
    pub fn generic(&'a self, reason: &'a Reason<'a>, name: &'a str) -> &'a Ty<'a> {
        // TODO (T69662957) must fill type args of Tgeneric
        self.mk(reason, Ty_::Tgeneric(self.alloc((name, &[][..]))))
    }
    pub fn fun(&'a self, reason: &'a Reason<'a>, ft: &'a FunType<'a>) -> &'a Ty<'a> {
        self.mk(reason, Ty_::Tfun(ft))
    }
    pub fn loclty(&'a self, ty: &'a Ty<'a>) -> InternalType<'a> {
        InternalType::LoclType(ty)
    }
    pub fn constraintty(&'a self, ty: ConstraintType<'a>) -> InternalType<'a> {
        InternalType::ConstraintType(ty)
    }
    pub fn funparam(&'a self, type_: &'a Ty<'a>) -> &'a FunParam<'a> {
        self.alloc(FunParam {
            type_: self.alloc(PossiblyEnforcedTy {
                type_,
                // TODO: set this correctly
                enforced: Enforcement::Unenforced,
            }),
            // TODO: set the following fields correctly
            flags: FunParamFlags::empty(),
            name: None,
            pos: Pos::none(),
        })
    }
    pub fn funtype(
        &'a self,
        params: &'a [&'a FunParam<'a>],
        implicit_params: &'a FunImplicitParams<'a>,
        ret: &'a Ty<'a>,
    ) -> &'a FunType<'a> {
        self.alloc(FunType {
            ret: self.alloc(PossiblyEnforcedTy {
                type_: ret,
                // TODO: set this correctly
                enforced: Enforcement::Unenforced,
            }),
            params,
            implicit_params,
            // TODO: set the following fields correctly
            arity: FunArity::Fstandard,
            tparams: &[],
            where_constraints: &[],
            flags: FunTypeFlags::empty(),
            ifc_decl: IfcFunDecl::FDPolicied(Some("#PUBLIC")),
        })
    }
}

/// Subtype props
impl<'a> TypeBuilder<'a> {
    pub fn conj(&'a self, v: BVec<'a, SubtypeProp<'a>>) -> SubtypeProp<'a> {
        self.alloc(SubtypePropEnum::Conj(v.into()))
    }
    pub fn disj(&'a self, v: BVec<'a, SubtypeProp<'a>>) -> SubtypeProp<'a> {
        self.alloc(SubtypePropEnum::Disj(v.into()))
    }
    pub fn is_subtype(&'a self, ty1: InternalType<'a>, ty2: InternalType<'a>) -> SubtypeProp<'a> {
        self.alloc(SubtypePropEnum::IsSubtype(ty1, ty2))
    }
    pub fn valid(&'a self) -> SubtypeProp<'a> {
        self.alloc(SubtypePropEnum::Conj(Vec::new_in(self.bumpalo)))
    }
    pub fn invalid(&'a self) -> SubtypeProp<'a> {
        self.alloc(SubtypePropEnum::Disj(Vec::new_in(self.bumpalo)))
    }
}

/// All reason builders go here
impl<'a> TypeBuilder<'a> {
    /// Make an Rnone reason. This does not belong here, but:
    pub fn mk_rnone(&'a self) -> &'a Reason<'a> {
        Reason::none()
    }

    pub fn mk_rinstantiate(
        &'a self,
        r0: &'a Reason<'a>,
        name: &'a str,
        r1: &'a Reason<'a>,
    ) -> &'a Reason<'a> {
        self.alloc(Reason::Rinstantiate(self.alloc((*r0, name, *r1))))
    }

    pub fn mk_rtype_variable_generics(
        &'a self,
        pos: &'a Pos,
        param_name: &'a str,
        fn_name: &'a str,
    ) -> &'a Reason<'a> {
        self.alloc(Reason::RtypeVariableGenerics(
            self.alloc((pos, param_name, fn_name)),
        ))
    }

    pub fn mk_rwitness(&'a self, pos: &'a Pos) -> &'a Reason<'a> {
        self.alloc(Reason::witness(pos))
    }
}

impl<'a> TypeBuilder<'a> {
    pub fn pos_none(&'a self) -> &'a Pos {
        Pos::none()
    }
}

impl<'a> TypeBuilder<'a> {
    pub fn env_with_self(&self) -> &mut ExpandEnv<'a> {
        // TODO(hrust) this_ty
        self.alloc(ExpandEnv {
            type_expansions: vec![in &self.bumpalo],
            substs: SMap::empty(),
        })
    }
}

#[macro_export]
macro_rules! avec {
    (in $bld:expr; $elem:expr; $n:expr) => (bumpalo::vec![in $bld.bumpalo(); $elem; $n]);
    (in $bld:expr) => (bumpalo::vec![in $bld.bumpalo()]);
    (in $bld:expr; $($x:expr),*) => (bumpalo::vec![in $bld.bumpalo(); $($x),*]);
    (in $bld:expr; $($x:expr,)*) => (bumpalo::vec![in $bld.bumpalo(); $($x),*]);
}

#[macro_export]
macro_rules! aset {
  ( ) => ({ Set::empty() });
  ( in $bld:expr; $($x:expr),* ) => ({
      let mut temp_map = Set::empty();
      $(
          temp_map = temp_map.add($bld, $x);
      )*
      temp_map
  });
}
